docs/capturing-arguments.md

```diff
@@ -0,0 +1,107 @@
+* use `Values` to access all the captured values so they can be asserted
+* (alternative flow) use `GetLastValue` to access the most recent captured value so it can be asserted
+
+???+ note "Values are only captured if the call matches the configuration"
+    When a call configuration includes one or more argument-capturing constraints, the argument
+    values are only captured if the call is triggered. If an incoming call does not match that
+    configured for the method or property, no arguments are captured.
+
+## Capturing and constraining at the same time
+
+Even though you can interrogate captured values after the fact, you may want to configure
+fake behavior to take effect only when incoming arguments meet a constraint. You can do this
+using syntax similar to that of non-capturing constraints.
+
+```csharp title="constrained capture" linenums="1" hl_lines="7 8"
+--8<--
+recipes/FakeItEasy.Recipes.CSharp/CapturingArguments.cs:ConstrainedCapture
+--8<--
+```
+
+## Capturing mutable arguments
+
+### The challenge
+
+Argument capture is shallow; there's no copying of object state.
+If a reference-based argument (e.g. a class instance, not a struct) is captured and
+subsequently modified by the test or production code, the updated state will be seen
+during the "assert" phase of the test.
+
+```csharp title="mutating captured values" linenums="1" hl_lines="13 14 15 21 22"
+--8<--
+recipes/FakeItEasy.Recipes.CSharp/CapturingArguments.cs:NaivelyCaptureMutatedList
+--8<--
+```
+
+Here a single list instance is twice passed to the production code, which removes
+the first element each time. The `CapturedArgument` captures a reference to the list each
+time, but does not preserve the internals of the list. So by the time the `Values` are examined,
+the list has had its first two elements removed, and this is reflected in the failing assertion.
+
+It's the same effect as if the following were run:
+
+```c#
+var list = new List<int> { 1, 2, 3 };
+var a = list;
+list.RemoveAt(0);
+var b = list;
+
+// both a and b point to a list with elements {2, 3}
+```
+
+### Capturing frozen state
+
+`CapturedArgument`s can be created with a transforming function (or "freezer") that runs on the
+argument value before saving it away. This behavior can be used to insulate the captured values
+from subsequent mutations.
+
+```csharp title="freezing state of captured values" linenums="1" hl_lines="2 3"
+--8<--
+recipes/FakeItEasy.Recipes.CSharp/CapturingArguments.cs:CaptureCopiedMutatedList
+--8<--
+```
+
+You can even transform values into a different type:
+
+```csharp title="freezing state of captured values as new type" linenums="1" hl_lines="2 3 19 20"
+--8<--
+recipes/FakeItEasy.Recipes.CSharp/CapturingArguments.cs:CaptureCopiedMutatedListToNewType
+--8<--
+```
+
+Aside from changing the freezer function, the `CapturedArgument` needs to have the typeparams
+supplied for the argument value and the captured value.

```

Comment:
I wonder if we could change the API a bit to take advantage of type inference to avoid having to specify all type params. Maybe something like this?
```csharp
var capturedLists = CapturedArgument
    .OfType<IEnumerable<string>>()
    .Freeze(l => string.Join(" - ", l!)
```